Don't treat x in pre-release and build metadata identifiers as a wild…

mamachanko opened the pull request:
…card

I don't think we would want to extend logic to wildcards _inside_ pre-release and build metadata identifiers.

Would need flow up through vendir into kapp-controller.

refs:
* https://github.com/carvel-dev/vendir/issues/123
* https://github.com/carvel-dev/kapp-controller/issues/1518

mamachanko commented:
In the absence of CI
```shell
pushd v4 && go test -v ./... && popd
```
passes ✅

Zebradil commented on v4/range.go:
```diff
@@ -381,6 +381,12 @@ func expandWildcardVersion(parts [][]string) ([][]string, error) {
 	return expandedParts, nil
 }
 
+// containsWildcard returns true iff there's a wildcard in any of the major, minor or patch components
```
```suggestion
// containsWildcard returns true if there's a wildcard in any of the major, minor or patch components
```

mamachanko replied:
I meant to express "if and only if" through the [conventional abbreviation _iff_](https://en.wikipedia.org/wiki/If_and_only_if).

Zebradil replied:
Some people might be not aware of this. I personally have seen usage of it just a couple of times over last 10 years. But it's up to you.

praveenrewar commented:
I am not an expert at semver, however the changes looks good to me. 
Out of curiosity, any reason not to use the changes from original [PR](https://github.com/blang/semver/pull/76/files) in blang/semver?

Zebradil commented:
In the [upstream PR](https://github.com/blang/semver/pull/76/files) this [test](https://github.com/carvel-dev/semver/pull/4/files#diff-c38d2fa72cd0dd2e7af1f424f5b61d97f0fafb6a97deafdeefd3c288d22031d7R391-R400) fails:

```
?       github.com/blang/semver/v4/examples [no test files]
--- FAIL: TestParseRange (0.00s)
    range_test.go:511: Invalid for case "1.2.3+experimental.2" matching "1.2.3+experimental.1": Expected false, got: true
    range_test.go:511: Invalid for case "1.2.3+experimental.2" matching "1.2.3+experimental.3": Expected false, got: true
FAIL
FAIL    github.com/blang/semver/v4  0.002s
FAIL
```

mamachanko commented:
> In the [upstream PR](https://github.com/blang/semver/pull/76/files) this [test](https://github.com/carvel-dev/semver/pull/4/files#diff-c38d2fa72cd0dd2e7af1f424f5b61d97f0fafb6a97deafdeefd3c288d22031d7R391-R400) fails:
> 
> ```
> ?       github.com/blang/semver/v4/examples [no test files]
> --- FAIL: TestParseRange (0.00s)
>     range_test.go:511: Invalid for case "1.2.3+experimental.2" matching "1.2.3+experimental.1": Expected false, got: true
>     range_test.go:511: Invalid for case "1.2.3+experimental.2" matching "1.2.3+experimental.3": Expected false, got: true
> FAIL
> FAIL    github.com/blang/semver/v4  0.002s
> FAIL
> ```

Thank you for bringing that up, @Zebradil! I wasn't aware of the pending (since 3y! 😆) https://github.com/blang/semver/pull/76. However, after studying it I partially incorporated the suggested changes here.

The test for `.x` as suggested by https://github.com/blang/semver/pull/76 is nice and simple, but the pre-release and build metadata identifiers need to be trimmed first. That's because either may contain `.x`. I've added tests for those scenarios and refactored things a little.

@Zebradil @praveenrewar ready for review! 🙇🏻 

Fwiw, I don't think we really consider github.com/blang/semver "upstream upstream" anymore since it is no longer maintained and github.com/carvel-dev/semver is ahead by one or two fixes and features. I may be wrong here.

Zebradil reviewed (approved):
LGTM

By the way, I think it worth considering switching to a more popular and better maintained semver library, for example this one: https://github.com/Masterminds/semver

joaopapereira reviewed (approved):
LGTM

joaopapereira commented:
@Zebradil I would tend to agree but we would have to ensure that the features that exist in this repo also exist in that semver library, so some work needs to be done to ensure it.

Zebradil commented:
> @Zebradil I would tend to agree but we would have to ensure that the features that exist in this repo also exist in that semver library, so some work needs to be done to ensure it.

Agree, it'd probably require a lot of effort without immediate benefits.
